extension/chrome/elements/compose-modules/formatters/encrypted-mail-msg-formatter.ts

```diff
@@ -77,26 +135,31 @@ export class EncryptedMsgMailFormatter extends BaseMailFormatter {
     const { bodyWithReplyToken, replyToken } = await this.getPwdMsgSendableBodyWithOnlineReplyMsgToken(authInfo, newMsg);
     const pgpMimeWithAttachments = await Mime.encode(bodyWithReplyToken, { Subject: newMsg.subject }, await this.view.attachmentsModule.attachment.collectAttachments());
     const { data: pwdEncryptedWithAttachments } = await this.encryptDataArmor(Buf.fromUtfStr(pgpMimeWithAttachments), newMsg.pwd, []); // encrypted only for pwd, not signed
-    const { url, externalId } = await this.view.acctServer.messageUpload(
+    return await this.view.acctServer.messageUpload(
       authInfo.uuid ? authInfo : undefined,
       pwdEncryptedWithAttachments,
       replyToken,
-      newMsg.from,
+      newMsg.from.email, // todo: Str.formatEmailWithOptionalName?
       newMsg.recipients,
       (p) => this.view.sendBtnModule.renderUploadProgress(p, 'FIRST-HALF'), // still need to upload to Gmail later, this request represents first half of progress
     );
-    return { url, externalId };
   };
 
-  private sendablePwdMsg = async (newMsg: NewMsgData, pubs: PubkeyResult[], { msgUrl, externalId }: { msgUrl: string, externalId?: string }, signingPrv?: Key) => {
+  private sendablePwdMsg = async (
+    newMsg: NewMsgData,
+    pubs: PubkeyResult[],
+    { msgUrl, externalId }: { msgUrl: string, externalId?: string },
+    attachments: Attachment[],
+    signingPrv?: Key) => {
     // encoded as: PGP/MIME-like structure but with attachments as external files due to email size limit (encrypted for pubkeys only)
     const msgBody = this.richtext ? { 'text/plain': newMsg.plaintext, 'text/html': newMsg.plainhtml } : { 'text/plain': newMsg.plaintext };
     const pgpMimeNoAttachments = await Mime.encode(msgBody, { Subject: newMsg.subject }, []); // no attachments, attached to email separately
     const { data: pubEncryptedNoAttachments } = await this.encryptDataArmor(Buf.fromUtfStr(pgpMimeNoAttachments), undefined, pubs, signingPrv); // encrypted only for pubs
-    const attachments = this.createPgpMimeAttachments(pubEncryptedNoAttachments).
-      concat(await this.view.attachmentsModule.attachment.collectEncryptAttachments(pubs)); // encrypted only for pubs
     const emailIntroAndLinkBody = await this.formatPwdEncryptedMsgBodyLink(msgUrl);
-    return await SendableMsg.createPwdMsg(this.acctEmail, this.headers(newMsg), emailIntroAndLinkBody, attachments, { isDraft: this.isDraft, externalId });
+    const headers = this.headers(newMsg);
+    return await SendableMsg.createPwdMsg(this.acctEmail, headers, emailIntroAndLinkBody,
+      this.createPgpMimeAttachments(pubEncryptedNoAttachments).concat(attachments),
+      { isDraft: this.isDraft, externalId });
```

Comment:
This is confusing - I imagine it may be wrong, but not sure.

When sending password-protected message, attachments should only be uploaded one time, to FES (or backend) as one big pgp/mime string that includes the message and the attachment.

That means that password-protected message that goes only to password recipients should no longer have any attachments on it at all.

Here, `sendablePwdMsg` seems to be adding attachments to each of these messages, and it's called in a loop so if I have 10 password recipints I may be uploading my attachments 10 times to google as implemented. I think? Which would be incorrect.

I didn't run it, just commenting from reading the code.

Reply:
`attachments` were probably `[]`, but, you're right, the code is not obvious. I refactored it considerably to be more straightforward